src/snowflake/cli/api/project/schemas/project_definition.py

```diff
@@ -105,8 +117,54 @@ def _convert_env(
 
 
 class DefinitionV20(_ProjectDefinitionBase):
-    entities: Dict = Field(
-        title="Entity definitions.",
+    entities: Dict[str, Annotated[Entity, Field(discriminator="type")]] = Field(
+        title="Entity definitions."
+    )
+
+    @model_validator(mode="before")
+    @classmethod
+    def apply_defaults(cls, data: Dict) -> Dict:
+        """
+        Applies default values that exist on the model but not specified in yml
+        """
+        if "defaults" in data and "entities" in data:
+            for key, entity in data["entities"].items():
+                entity_type = entity["type"]
+                if entity_type not in v2_entity_types_map:
+                    continue
+                entity_model = v2_entity_types_map[entity_type]
+                for default_key, default_value in data["defaults"].items():
+                    if (
+                        default_key in entity_model.model_fields
+                        and default_key not in entity
+                    ):
+                        entity[default_key] = default_value
+        return data
+
+    @field_validator("entities", mode="after")
+    @classmethod
+    def validate_entities(cls, entities: Dict[str, Entity]) -> Dict[str, Entity]:
+        for key, entity in entities.items():
+            # TODO Automatically detect TargetFields to validate
+            if entity.type == ApplicationEntity.get_type():
+                if isinstance(entity.from_.target, TargetField):
+                    target = str(entity.from_.target)
+                    if target not in entities:
+                        raise ValueError(f"No such target: {target}")
+                    else:
+                        # Validate the target type
+                        target_cls = entity.from_.__class__.model_fields["target"]
+                        target_type = target_cls.annotation.__args__[0]
+                        actual_target_type = entities[target].__class__
+                        if target_type and target_type is not actual_target_type:
+                            raise ValueError(
+                                f"Target type mismatch. Expected {target_type.__name__}, got {actual_target_type.__name__}"
```

Comment:
I imagine that there will more checks on the enitites should we consider something like
```suggestion
            # TODO Automatically detect TargetFields to validate
            if entity.type == ApplicationEntity.get_type():
               self._validate_application_targets(entity)
```
WDYT?

Reply:
I split it into 2 methods to have slightly better separation, but we will need to refactor deeper when we find a generic way for this validation. Will address it in a dedicated PR.